### DISCUSSION TITLE
Please tell me how to restore scrolling when using measureElement.

### DISCUSSION THREAD

**Discussion by shunshimono** [TanStack/virtual#731] [Q&A] [answered]
### Reference Discussion
https://github.com/TanStack/virtual/issues/378

### Issue
Restoring the scroll position during the initial rendering using the value saved in session storage

### Expected Behavior
The scroll position should be restored when returning to the page

### Confirmed
- [x] Values are saved in storage
- [x]  The correct value is returned to initialOffset

### My Code

```tsx
"use client";

import {
  useWindowVirtualizer,
  observeWindowOffset,
  measureElement as defaultMeasureElement,
} from "@tanstack/react-virtual";
import { useRef } from "react";
import Link from "next/link";

import { faker } from "@faker-js/faker";

const randomNumber = (min: number, max: number) =>
  faker.datatype.number({ min, max });

const sentences = new Array(100)
  .fill(true)
  .map(() => faker.lorem.sentence(randomNumber(20, 70)));

export default function Page() {
  const listRef = useRef<HTMLDivElement | null>(null);
  const cachedSizesRef = useRef<Map<number, number>>(new Map());

  const measureElement = (
    element: Element,
    entry: ResizeObserverEntry | undefined,
    instance: any,
  ) => {
    const index = instance.indexFromElement(element);
    const size = defaultMeasureElement(element, entry, instance);
    cachedSizesRef.current.set(index, size);
    return size;
  };

  const virtualizer = useWindowVirtualizer({
    count: 100,
    estimateSize: () => 35,
    overscan: 5,
    measureElement,
    initialOffset: (() => {
      if (typeof sessionStorage !== "undefined") {
        return parseInt(
          sessionStorage.getItem("virtualizer_scrollOffset") || "",
        );
      }
      return 0;
    })(),
    observeElementOffset: (instance, cb) => {
      return observeWindowOffset(instance, (offset) => {
        if (typeof sessionStorage !== "undefined") {
          sessionStorage.setItem(
            "virtualizer_scrollOffset",
            offset?.toString(),
          );
        }
        cb(offset, instance.isScrolling);
      });
    },
  });

  return (
    <>
      <div ref={listRef} className="List">
        <div
          style={{
            height: `${virtualizer.getTotalSize()}px`,
            width: "100%",
            position: "relative",
          }}
        >
          {virtualizer.getVirtualItems().map((item) => (
            <div
              key={item.key}
              ref={virtualizer.measureElement}
              data-index={item.index}
              className={item.index % 2 ? "ListItemOdd" : "ListItemEven"}
              style={{
                position: "absolute",
                top: 0,
                left: 0,
                width: "100%",
                transform: `translateY(${item.start}px)`,
              }}
            >
              <h1>Row {item.index}</h1>
              <Link href={"/about"}>about</Link>
              <p>{sentences[item.index]}</p>
            </div>
          ))}
        </div>
      </div>
    </>
  );
}

```

### Environment
[link](https://codesandbox.io/p/devbox/virtualwindowlist-g99vpm?file=%2Fapp%2Fpage.tsx%3A35%2C16&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clw9z35k60007356jp4dtjvhs%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clw9z35k60002356jy6rhlqqa%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clw9z35k60004356jzvgono5h%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clw9z35k60006356jltb3a483%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B57.404919156249065%252C42.595080843750935%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clw9z35k60002356jy6rhlqqa%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clw9z35k60001356jkf1sc8nn%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522filepath%2522%253A%2522%252FREADME.md%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%252C%257B%2522id%2522%253A%2522clwa06gho0002356jxeoku84m%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A35%252C%2522startColumn%2522%253A16%252C%2522endLineNumber%2522%253A35%252C%2522endColumn%2522%253A16%257D%255D%252C%2522filepath%2522%253A%2522%252Fapp%252Fpage.tsx%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clw9z35k60002356jy6rhlqqa%2522%252C%2522activeTabId%2522%253A%2522clwa06gho0002356jxeoku84m%2522%257D%252C%2522clw9z35k60006356jltb3a483%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clw9z35k60005356jmobji7zk%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_PORT%2522%252C%2522taskId%2522%253A%2522dev%2522%252C%2522port%2522%253A3000%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clw9z35k60006356jltb3a483%2522%252C%2522activeTabId%2522%253A%2522clw9z35k60005356jmobji7zk%2522%257D%252C%2522clw9z35k60004356jzvgono5h%2522%253A%257B%2522id%2522%253A%2522clw9z35k60004356jzvgono5h%2522%252C%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clw9z35k60003356jznt6cg1d%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522TASK_LOG%2522%252C%2522taskId%2522%253A%2522dev%2522%257D%255D%252C%2522activeTabId%2522%253A%2522clw9z35k60003356jznt6cg1d%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A20.9375%257D)

**Comment by shunshimono** [comment]
I've found that while the scroll position is set to the correct value, the elements that were originally at that scroll position are not being displayed when restored.

**Reply by piecyk** [reply]
Yes, as you need store also `measurementsCache` and pass it to `initialMeasurementsCache`

**Reply by shunshimono** [reply, marked as answer]
@piecyk 
Thank you. I was able to resolve it by saving measurementsCache to storage when leaving the page, and setting that value to initialMeasurementsCache when remounting.


### Resumable code
```tsx
"use client";

import { useWindowVirtualizer } from "@tanstack/react-virtual";
import { useRef, useEffect } from "react";
import Link from "next/link";

import { faker } from "@faker-js/faker";

const randomNumber = (min: number, max: number) =>
  faker.datatype.number({ min, max });

const sentences = new Array(100)
  .fill(true)
  .map(() => faker.lorem.sentence(randomNumber(20, 70)));

export default function Page() {
  const listRef = useRef<HTMLDivElement | null>(null);
  const virtualizer = useWindowVirtualizer({
    count: 100,
    estimateSize: () => 35,
    overscan: 5,
    initialOffset: (() => {
      if (typeof sessionStorage !== "undefined") {
        return parseInt(
          sessionStorage.getItem("virtualizer_scrollOffset") || "",
        );
      }
      return 0;
    })(),
    initialMeasurementsCache: (() => {
      if (typeof sessionStorage === "undefined") return;
      const savedCache = sessionStorage.getItem(
        "virtualizer_measurementsCache",
      );
      return savedCache ? JSON.parse(savedCache) : undefined;
    })(),
  });

  useEffect(() => {
    sessionStorage.setItem(
      "virtualizer_measurementsCache",
      JSON.stringify(Array.from(virtualizer.measurementsCache)),
    );
  }, [virtualizer.measurementsCache]);

  return (
      <div ref={listRef} className="List">
        <div
          style={{
            height: `${virtualizer.getTotalSize()}px`,
            width: "100%",
            position: "relative",
          }}
        >
          {virtualizer.getVirtualItems().map((item) => (
            <div
              key={item.key}
              ref={virtualizer.measureElement}
              data-index={item.index}
              className={item.index % 2 ? "ListItemOdd" : "ListItemEven"}
              style={{
                position: "absolute",
                top: 0,
                left: 0,
                width: "100%",
                transform: `translateY(${item.start}px)`,
              }}
            >
              <h1>Row {item.index}</h1>
              <Link
                href={"/about"}
                onClick={() => {
                  sessionStorage.setItem(
                    "virtualizer_scrollOffset",
                    document.documentElement.scrollTop?.toString(),
                  );
                }}
              >
                about
              </Link>
              <p>{sentences[item.index]}</p>
            </div>
          ))}
        </div>
      </div>
  );
}
```

**Reply by DMaxence** [reply]
@shunshimono thank you!
